### PR TITLE
Gradient shenanigans

### PR DESCRIPTION
**Pull request by pmichel31415**
This integrates my code for combining gradients for multi objective training. Needs testing.

### CODE REVIEW

File: fairseq/optim/multiobj/multiobj_sgd.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+                param_state = self.state[p]
+                # skip frozen parameters (TODO: remove this)
+                if getattr(param_state, "frozen", False):
+                    continue
+                # Actually save the gradient
+                param_state["aux_grad"] = th.zeros_like(p.data)
+                if p.grad is None:
+                    continue
+                d_p = p.grad.data
+                param_state["aux_grad"].add_(d_p)
```
Comment (xianxl) [inline fairseq/optim/multiobj/multiobj_sgd.py]: Is this is the case that there's alway one auxiliary gradient?

File: multiobj_train.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+    # Initialize data iterator
+    itr = epoch_itr.next_epoch_itr(
+        fix_batches_to_gpus=args.fix_batches_to_gpus)
+    itr = iterators.GroupedIterator(itr, update_freq)
+    progress = progress_bar.build_progress_bar(
+        args, itr, epoch_itr.epoch, no_progress_bar='simple',
+    )
+
+    # Auxiliary iterator
```
Comment (xianxl) [inline multiobj_train.py]: I wonder whether we need a separate iterator, vs. a single iterator on a multi-dataset, e.g. https://github.com/pytorch/fairseq/blob/master/fairseq/data/round_robin_zip_datasets.py.

**Review by xianxl — APPROVED** [review]

File: fairseq/optim/multiobj/multiobj_sgd.py (hunk 1 above)
Comment (pmichel31415) [reply fairseq/optim/multiobj/multiobj_sgd.py]: For now yes. I am planning to add support for multiple auxiliary gradients later but this will complicate the code so I'm leaving it for later.

File: multiobj_train.py (hunk 2 above)
Comment (pmichel31415) [reply multiobj_train.py]: Good point. To be honest I used 2 iterators because I didn't want to write another Dataset class. I don't think RoundRobin fits the bill completely since it matches the length of the longest dataset. Not sure wheter one solution is better than the other tbh

Comment (pmichel31415) [reply multiobj_train.py]: Actually the current version probably gives greater control over the batch size of each data source so it's probably better

**Comment by pmichel31415** [comment]
Feel free to merge.
